File: php/ant_ffi/src/Types/Scratchpad.php
<?php

declare(strict_types=1);

namespace AntFfi\Types;

use AntFfi\FFILoader;
use AntFfi\NativeHandle;
use AntFfi\RustBuffer;
use FFI;
use FFI\CData;

/**
 * Address of a scratchpad on the network.
 */
final class ScratchpadAddress extends NativeHandle
{
    /**
     * Create a scratchpad address from the owner public key.
     */
    public static function fromOwner(PublicKey $owner): self
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');

        $handle = $ffi->uniffi_ant_ffi_fn_constructor_scratchpadaddress_new(
            $owner->cloneForCall(),
            FFI::addr($status)
        );

        RustBuffer::checkStatus($status);

        return new self($handle);
    }

    /**
     * Create from a hex string.
     */
    public static function fromHex(string $hex): self
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');

        $hexBuffer = RustBuffer::fromString($hex);
        $handle = $ffi->uniffi_ant_ffi_fn_constructor_scratchpadaddress_from_hex(
            $hexBuffer,
            FFI::addr($status)
        );

        RustBuffer::checkStatus($status);

        return new self($handle);
    }

    /**
     * Get the owner public key.
     */
    public function owner(): PublicKey
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');

        $handle = $ffi->uniffi_ant_ffi_fn_method_scratchpadaddress_owner(
            $this->cloneForCall(),
            FFI::addr($status)
        );

        RustBuffer::checkStatus($status);

        return new PublicKey($handle);
    }

    /**
     * Convert to hex string.
     */
    public function toHex(): string
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');
        $resultBuffer = $ffi->new('RustBuffer');

        $ffi->uniffi_ant_ffi_fn_method_scratchpadaddress_to_hex(
            FFI::addr($resultBuffer),
            $this->cloneForCall(),
            FFI::addr($status)
        );

        RustBuffer::checkStatus($status);

        $result = RustBuffer::toString($resultBuffer);
        RustBuffer::free($resultBuffer);

        return $result;
    }

    protected function freeHandle(): void
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');
        $ffi->uniffi_ant_ffi_fn_free_scratchpadaddress($this->handle, FFI::addr($status));
    }

    protected function cloneHandle(): CData
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');
        return $ffi->uniffi_ant_ffi_fn_clone_scratchpadaddress($this->handle, FFI::addr($status));
    }
}

/**
 * A mutable scratchpad on the network.
 *
 * Scratchpads can be updated by the owner and are useful for storing
 * frequently changing data.
 */
final class Scratchpad extends NativeHandle
{
    /**
     * Create a new scratchpad.
     *
     * @param SecretKey $owner The owner's secret key
     * @param int $contentType Content type identifier
     * @param string $data The data to store
     * @param int $counter Monotonically increasing counter
     */
    public static function create(SecretKey $owner, int $contentType, string $data, int $counter): self
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');

        $dataBuffer = RustBuffer::fromStringWithPrefix($data);
        $handle = $ffi->uniffi_ant_ffi_fn_constructor_scratchpad_new(
            $owner->cloneForCall(),
            $contentType,
            $dataBuffer,
            $counter,
            FFI::addr($status)
        );

        RustBuffer::checkStatus($status);

        return new self($handle);
    }

    /**
     * Get the address of this scratchpad.
     */
    public function address(): ScratchpadAddress
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');

        $handle = $ffi->uniffi_ant_ffi_fn_method_scratchpad_address(
            $this->cloneForCall(),
            FFI::addr($status)
        );

        RustBuffer::checkStatus($status);

        return new ScratchpadAddress($handle);
    }

    /**
     * Get the counter value.
     */
    public function counter(): int
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');

        $result = $ffi->uniffi_ant_ffi_fn_method_scratchpad_counter(
            $this->cloneForCall(),
            FFI::addr($status)
        );

        RustBuffer::checkStatus($status);

        return $result;
    }

    /**
     * Get the data encoding type.
     */
    public function dataEncoding(): int
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');

        $result = $ffi->uniffi_ant_ffi_fn_method_scratchpad_data_encoding(
            $this->cloneForCall(),
            FFI::addr($status)
        );

        RustBuffer::checkStatus($status);

        return $result;
    }

    /**
     * Alias for dataEncoding() for backwards compatibility.
     */
    public function contentType(): int
    {
        return $this->dataEncoding();
    }

    /**
     * Get the encrypted data.
     */
    public function encryptedData(): string
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');
        $resultBuffer = $ffi->new('RustBuffer');

        $ffi->uniffi_ant_ffi_fn_method_scratchpad_encrypted_data(
            FFI::addr($resultBuffer),
            $this->cloneForCall(),
            FFI::addr($status)
        );

        RustBuffer::checkStatus($status);

        $result = RustBuffer::toString($resultBuffer);
        RustBuffer::free($resultBuffer);

        return $result;
    }

    /**
     * Alias for encryptedData() for backwards compatibility.
     */
    public function data(): string
    {
        return $this->encryptedData();
    }

    /**
     * Decrypt the scratchpad data with the owner's secret key.
     */
    public function decryptData(SecretKey $sk): string
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');
        $resultBuffer = $ffi->new('RustBuffer');

        $ffi->uniffi_ant_ffi_fn_method_scratchpad_decrypt_data(
            FFI::addr($resultBuffer),
            $this->cloneForCall(),
            $sk->cloneForCall(),
            FFI::addr($status)
        );

        RustBuffer::checkStatus($status);

        $result = RustBuffer::toString($resultBuffer);
        RustBuffer::free($resultBuffer);

        return $result;
    }

    /**
     * Verify the scratchpad signature is valid.
     * Throws an exception if invalid.
     */
    public function verify(): void
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');

        $ffi->uniffi_ant_ffi_fn_func_scratchpad_verify(
            $this->cloneForCall(),
            FFI::addr($status)
        );

        RustBuffer::checkStatus($status);
    }

    /**
     * Check if the scratchpad signature is valid.
     */
    public function isValid(): bool
    {
        try {
            $this->verify();
            return true;
        } catch (\Exception $e) {
            return false;
        }
    }

    /**
     * Get the owner's public key.
     */
    public function owner(): PublicKey
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');

        $handle = $ffi->uniffi_ant_ffi_fn_method_scratchpad_owner(
            $this->cloneForCall(),
            FFI::addr($status)
        );

        RustBuffer::checkStatus($status);

        return new PublicKey($handle);
    }

    protected function freeHandle(): void
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');
        $ffi->uniffi_ant_ffi_fn_free_scratchpad($this->handle, FFI::addr($status));
    }

    protected function cloneHandle(): CData
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');
        return $ffi->uniffi_ant_ffi_fn_clone_scratchpad($this->handle, FFI::addr($status));
    }
}

File: dart/ant_ffi/lib/src/types/scratchpad.dart
import 'dart:ffi';
import 'dart:typed_data';
import 'package:ffi/ffi.dart';
import '../native/bindings.dart';
import '../native/library.dart';
import '../native/rust_buffer.dart';
import 'public_key.dart';
import 'secret_key.dart';

late final _bindings = AntFfiBindings(antFfiLib);

void _checkStatus(RustCallStatus status, String operation) {
  if (status.code != 0) {
    String errorMessage = '$operation failed with code ${status.code}';
    if (status.errorBuf.len > 0) {
      try {
        errorMessage = rustBufferToStringWithPrefix(status.errorBuf);
      } catch (_) {
        try {
          errorMessage = rustBufferToString(status.errorBuf);
        } catch (_) {}
      }
    }
    throw AntFfiException(errorMessage, status.code);
  }
}

/// Address of a scratchpad on the network.
class ScratchpadAddress {
  final Pointer<Void> _handle;
  bool _disposed = false;

  ScratchpadAddress._(this._handle);

  /// Creates a ScratchpadAddress from a raw handle.
  factory ScratchpadAddress.fromHandle(Pointer<Void> handle) {
    return ScratchpadAddress._(handle);
  }

  Pointer<Void> _clone() {
    _checkNotDisposed();
    final status = calloc<RustCallStatus>();
    try {
      final cloned = _bindings.uniffi_ant_ffi_fn_clone_scratchpadaddress(_handle, status);
      _checkStatus(status.ref, 'ScratchpadAddress.clone');
      return cloned;
    } finally {
      calloc.free(status);
    }
  }

  /// Returns a cloned handle for use by client operations.
  Pointer<Void> cloneHandle() => _clone();

  /// Creates a ScratchpadAddress from a public key.
  factory ScratchpadAddress.fromPublicKey(PublicKey publicKey) {
    final status = calloc<RustCallStatus>();
    try {
      final handle = _bindings.uniffi_ant_ffi_fn_constructor_scratchpadaddress_new(
        publicKey.cloneHandle(),
        status,
      );
      _checkStatus(status.ref, 'ScratchpadAddress.fromPublicKey');
      return ScratchpadAddress._(handle);
    } finally {
      calloc.free(status);
    }
  }

  /// Creates a ScratchpadAddress from a hex-encoded string.
  factory ScratchpadAddress.fromHex(String hex) {
    final status = calloc<RustCallStatus>();
    try {
      final hexBuffer = stringToRustBuffer(hex);
      final handle = _bindings.uniffi_ant_ffi_fn_constructor_scratchpadaddress_from_hex(hexBuffer, status);
      _checkStatus(status.ref, 'ScratchpadAddress.fromHex');
      return ScratchpadAddress._(handle);
    } finally {
      calloc.free(status);
    }
  }

  /// Returns the owner public key of this scratchpad address.
  PublicKey owner() {
    _checkNotDisposed();
    final status = calloc<RustCallStatus>();
    try {
      final handle = _bindings.uniffi_ant_ffi_fn_method_scratchpadaddress_owner(_clone(), status);
      _checkStatus(status.ref, 'ScratchpadAddress.owner');
      return PublicKey.fromHandle(handle);
    } finally {
      calloc.free(status);
    }
  }

  /// Converts the address to a hex-encoded string.
  String toHex() {
    _checkNotDisposed();
    final status = calloc<RustCallStatus>();
    try {
      final buffer = _bindings.uniffi_ant_ffi_fn_method_scratchpadaddress_to_hex(_clone(), status);
      _checkStatus(status.ref, 'ScratchpadAddress.toHex');
      final result = rustBufferToString(buffer);
      buffer.free();
      return result;
    } finally {
      calloc.free(status);
    }
  }

  /// Releases the native resources.
  void dispose() {
    if (_disposed) return;
    _disposed = true;
    final status = calloc<RustCallStatus>();
    try {
      _bindings.uniffi_ant_ffi_fn_free_scratchpadaddress(_handle, status);
    } finally {
      calloc.free(status);
    }
  }

  void _checkNotDisposed() {
    if (_disposed) {
      throw StateError('ScratchpadAddress has been disposed');
    }
  }
}

/// A mutable scratchpad for storing temporary data on the network.
class Scratchpad {
  final Pointer<Void> _handle;
  bool _disposed = false;

  Scratchpad._(this._handle);

  /// Creates a Scratchpad from a raw handle.
  factory Scratchpad.fromHandle(Pointer<Void> handle) {
    return Scratchpad._(handle);
  }

  Pointer<Void> _clone() {
    _checkNotDisposed();
    final status = calloc<RustCallStatus>();
    try {
      final cloned = _bindings.uniffi_ant_ffi_fn_clone_scratchpad(_handle, status);
      _checkStatus(status.ref, 'Scratchpad.clone');
      return cloned;
    } finally {
      calloc.free(status);
    }
  }

  /// Returns a cloned handle for use by client operations.
  Pointer<Void> cloneHandle() => _clone();

  /// Creates a new Scratchpad.
  factory Scratchpad.create(SecretKey owner, int dataEncoding, Uint8List data, int counter) {
    final status = calloc<RustCallStatus>();
    try {
      final dataBuffer = uint8ListToRustBuffer(data);
      final handle = _bindings.uniffi_ant_ffi_fn_constructor_scratchpad_new(
        owner.cloneHandle(),
        dataEncoding,
        dataBuffer,
        counter,
        status,
      );
      _checkStatus(status.ref, 'Scratchpad.create');
      return Scratchpad._(handle);
    } finally {
      calloc.free(status);
    }
  }

  /// Returns the address of this scratchpad.
  ScratchpadAddress address() {
    _checkNotDisposed();
    final status = calloc<RustCallStatus>();
    try {
      final handle = _bindings.uniffi_ant_ffi_fn_method_scratchpad_address(_clone(), status);
      _checkStatus(status.ref, 'Scratchpad.address');
      return ScratchpadAddress.fromHandle(handle);
    } finally {
      calloc.free(status);
    }
  }

  /// Returns the data encoding type.
  int dataEncoding() {
    _checkNotDisposed();
    final status = calloc<RustCallStatus>();
    try {
      final result = _bindings.uniffi_ant_ffi_fn_method_scratchpad_data_encoding(_clone(), status);
      _checkStatus(status.ref, 'Scratchpad.dataEncoding');
      return result;
    } finally {
      calloc.free(status);
    }
  }

  /// Returns the counter value.
  int counter() {
    _checkNotDisposed();
    final status = calloc<RustCallStatus>();
    try {
      final result = _bindings.uniffi_ant_ffi_fn_method_scratchpad_counter(_clone(), status);
      _checkStatus(status.ref, 'Scratchpad.counter');
      return result;
    } finally {
      calloc.free(status);
    }
  }

  /// Decrypts and returns the data using the owner's secret key.
  Uint8List decryptData(SecretKey secretKey) {
    _checkNotDisposed();
    final status = calloc<RustCallStatus>();
    try {
      final buffer = _bindings.uniffi_ant_ffi_fn_method_scratchpad_decrypt_data(
        _clone(),
        secretKey.cloneHandle(),
        status,
      );
      _checkStatus(status.ref, 'Scratchpad.decryptData');
      final result = rustBufferToUint8ListWithPrefix(buffer);
      buffer.free();
      return result;
    } finally {
      calloc.free(status);
    }
  }

  /// Returns the owner public key.
  PublicKey owner() {
    _checkNotDisposed();
    final status = calloc<RustCallStatus>();
    try {
      final handle = _bindings.uniffi_ant_ffi_fn_method_scratchpad_owner(_clone(), status);
      _checkStatus(status.ref, 'Scratchpad.owner');
      return PublicKey.fromHandle(handle);
    } finally {
      calloc.free(status);
    }
  }

  /// Returns the encrypted data.
  Uint8List encryptedData() {
    _checkNotDisposed();
    final status = calloc<RustCallStatus>();
    try {
      final buffer = _bindings.uniffi_ant_ffi_fn_method_scratchpad_encrypted_data(_clone(), status);
      _checkStatus(status.ref, 'Scratchpad.encryptedData');
      final result = rustBufferToUint8ListWithPrefix(buffer);
      buffer.free();
      return result;
    } finally {
      calloc.free(status);
    }
  }

  /// Releases the native resources.
  void dispose() {
    if (_disposed) return;
    _disposed = true;
    final status = calloc<RustCallStatus>();
    try {
      _bindings.uniffi_ant_ffi_fn_free_scratchpad(_handle, status);
    } finally {
      calloc.free(status);
    }
  }

  void _checkNotDisposed() {
    if (_disposed) {
      throw StateError('Scratchpad has been disposed');
    }
  }
}
